Automate releases on GitHub

MRDGH2821 opened the pull request:
This automates generation of blueprint release by using GitHub Actions feature.

This is triggered only when a git tag with proper folder name is created.
For example:

Suppose you want to now create a new release for self building factory v6.0.2

- Git tag should be `self-building-factory-6.0.2`.
- This workflow will then search for `self-building-factory` folder & build (encode) the blueprint using fatul.
- Encoded blueprint file name will be `self-building-fatory-6.0.2.txt`
- Once encoding/building is done, it will automatically create a github release with the name `Self Building Factory 6.0.2` & attach the blueprint file

MRDGH2821 commented:
Here is how it will look

![image](https://github.com/user-attachments/assets/03be1222-acf4-4bb3-8778-76c9f42b37b0)
![image](https://github.com/user-attachments/assets/3f282ef0-9d5f-4a29-a0e0-81a9862d15ee)

bcwhite-code commented on .github/workflows/release-the-blueprint.yaml:
```diff
@@ -0,0 +1,42 @@
+          version=${GITHUB_REF##*/}
+          version=${version//[!0-9.]/}
+
+          release_name=$(echo "${folder//-/ }" | sed -e 's/\b\(.\)/\u\1/g')
```
`sed -e "s/\b\(.\)/\u\1/g;s/Brians/Brian's/g"`

or even

`echo "$folder" | sed -e "s/-/ /g;s/\b\(.\)/\u\1/g;s/Brians/Brian's/g"`

MRDGH2821 replied:
Resolved in https://github.com/bcwhite-code/brians-blueprints/pull/27/commits/42a4b177579ec1110a3dc9d9f57795408c057dc4

MRDGH2821 commented:
I have added the new regex.
Along with that, I modified the workflow to include latest changelog whenever a new release gets published.

Additionally I fixed the formatting to one Changelog file & moved json files of `brians-train-auxilary` into `book` subfolder.

This should be now good to go 🤔

bcwhite-code commented on tools/extract-changelog.py:
```diff
@@ -0,0 +1,47 @@
+
+
+if __name__ == "__main__":
+    if len(sys.argv) != 2:
```
Take a second parameter that is the expected version number and only return that changelog.  That way, if the file isn't updated, it won't erroneously include the changes for the previous version.

bcwhite-code commented on tools/extract-changelog.py:
```diff
@@ -0,0 +1,47 @@
+
+
+if __name__ == "__main__":
+    if len(sys.argv) != 2:
```
Take a second parameter that is the expected version number and only return if the latest changelog matches that version.  That way, if the file isn't updated, it won't erroneously include the changes for the previous version.

Or enhance it to return the changelog for _any_ specific version.

MRDGH2821 replied:
Resolved in https://github.com/bcwhite-code/brians-blueprints/pull/27/commits/d372ceb88d47dc84c0a9a5d0a4be54295a47d8f3

bcwhite-code commented on tools/extract-changelog.py:
```diff
@@ -0,0 +1,49 @@
+    if version_header:
+        version_header = version_header.group()
+        # Find the section corresponding to the version
+        section = re.search(
```
Since the version-header search above is now a complete line-match, extract the body in the same regex instead of doing the search twice.

MRDGH2821 replied:
The version number is derived from git tag, for eg:
If the git tag is `"self-building-factory-5.0.0"`,
`folder=self-building-factory` & `version=5.0.0`.

This is getting extracted from `.github/workflows/release-the-blueprint.yaml#L20` and passed by `.github/workflows/release-the-blueprint.yaml#L36`.

`CHANGELOG.md` has headers with markdown h2, i.e., `## 5.0.0`, so regex match is needed still.

You may use this following bash script to test things out.

```bash
#!/usr/bin/env bash

GITHUB_REF="self-building-factory-5.0.0"
folder=${GITHUB_REF#*/tags/}
folder=${folder%-*}
version=${GITHUB_REF##*/}
version=${version//[!0-9.]/}

# release_name=$(echo "${folder//-/ }" | sed -e 's/\b\(.\)/\u\1/g')
release_name=$(echo "${folder//-/ }" | sed -e "s/\b\(.\)/\u\1/g;s/Brians/Brian's/g")
release_name="$release_name $version"

echo "folder=$folder"
echo "version=$version"
echo "release_name=$release_name"
```

bcwhite-code replied:
I understand that the regex is needed but a single regex can find the desired header **and** extract the body.  You shouldn't need to do the regex match twice.
